src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs

```diff
@@ -747,8 +747,22 @@ public BoundExpression Sequence(params BoundExpression[] parts)
         public BoundExpression Sequence(ImmutableArray<LocalSymbol> locals, params BoundExpression[] parts)
         {
             var builder = ArrayBuilder<BoundExpression>.GetInstance();
-            for (int i = 0; i < parts.Length - 1; i++) builder.Add(parts[i]);
+            for (int i = 0; i < parts.Length - 1; i++)
+            {
+                var part = parts[i];
+                if (LocalRewriter.ReadIsSideeffecting(part))
```

Comment:
I think adjusting the name of the API is the best option here. This factory was designed to be used for the purpose of lowering, so I don't think there is any need for the version without the optimization.

---
In reply to: [77089986](https://github.com/dotnet/roslyn/pull/13501#discussion_r77089986) [](ancestors = 77089986,77089404)